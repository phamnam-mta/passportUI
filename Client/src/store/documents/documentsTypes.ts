export enum DocumentStatus {
    Loading = "Loading",
    Loaded = "Loaded",
    Analyzing = "Analyzing",
    Analyzed = "Analyzed",
    AnalyzeFailed = "AnalyzeFailed",
    Labeled = "Labeled",
}

export interface IDocumentStates {
    loadingStatus: DocumentStatus.Loaded | DocumentStatus.Loading;
    analyzingStatus?: DocumentStatus.Analyzed | DocumentStatus.Analyzing | DocumentStatus.AnalyzeFailed;
    labelingStatus?: DocumentStatus.Labeled;
}

export interface IRawDocument {
    name: string;
    type: string;
    url: string;
}

export interface IDocument extends IRawDocument {
    thumbnail: string;
    numPages: number;
    currentPage: number;
    states: IDocumentStates;
    expirationTime?: number;
}
