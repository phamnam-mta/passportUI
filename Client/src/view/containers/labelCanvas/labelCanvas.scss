@import "~@fluentui/react/dist/sass/References";

.label-canvas {
    display: flex;
    flex-grow: 1;
    flex-direction: column;

    .label-canvas-command-bar {
        display: flex;
        justify-content: space-between;
        align-items: center;
        height: 44px;
        padding: 0 24px;
        border-bottom: 1px solid $ms-color-gray30;
    }

    .label-canvas-control-bar {
        display: grid;
        grid-template-columns: 1fr auto 1fr;
        align-items: center;
        height: 44px;
        border-top: 1px solid $ms-color-gray30;
        padding: 0 24px;

        .label-canvas-page-control {
            grid-column-start: 2;
        }

        .label-canvas-tool-bar {
            display: flex;
            margin-left: auto;
        }
    }

    .label-canvas-image-map {
        display: flex;
        flex-grow: 1;
        position: relative;
        overflow: hidden;

        .label-canvas-overlay {
            display: flex;
            flex-direction: column;
            align-items: center;
            position: absolute;
            background-color: rgba(255, 255, 255, 0.9);
            height: 100%;
            width: 100%;
        }
    }
}
