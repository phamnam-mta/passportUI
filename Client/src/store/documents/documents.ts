import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { resetPredictions } from "store/predictions/predictions";
import { DocumentLoaderFactory, IDocumentLoader } from "utils/documentLoader";
import { DocumentStatus, IDocument, IRawDocument } from "./documentsTypes";

export type DocumentsState = { documents: IDocument[]; currentDocument: IDocument | null };

export const initialState: DocumentsState = { documents: [], currentDocument: null };

export const addDocuments = createAsyncThunk("documents/addDocuments", async (docsToAdd: IRawDocument[]) => {
    return Promise.all(docsToAdd.map(async (doc) => (await getLoader(doc)).loadDocumentMeta()));
});

export const setCurrentDocument = createAsyncThunk("documents/setCurrentDocument", async (document: IDocument) => {
    const loader = await getLoader(document);
    const documentPage = await loader.loadDocumentPage(document.currentPage);
    return { document, documentPage };
});

export const setCurrentPage = createAsyncThunk("documents/setCurrentPage", async (pageNumber: number, thunkApi) => {
    const { documents: state } = thunkApi.getState() as any;
    const loader = await getLoader(state.currentDocument);
    const documentPage = await loader.loadDocumentPage(pageNumber);
    return { pageNumber, documentPage };
});

const documentLoaders = new Map<string, IDocumentLoader>();

const getLoader = async (document: IRawDocument): Promise<IDocumentLoader> => {
    let loader = documentLoaders.get(document.url);
    if (!loader) {
        loader = await DocumentLoaderFactory.makeLoader(document);
        documentLoaders.set(document.url, loader);
    }

    return loader;
};

const documentsSlice = createSlice({
    name: "documents",
    initialState,
    reducers: {
        deleteDocument(state, action) {
            const docNameToDelete = action.payload.name;
            state.documents = state.documents.filter((document) => document.name !== docNameToDelete);
            documentLoaders.delete(action.payload.url);
        },
        setDocumentAnalyzingStatus(state, action) {
            const { name, status } = action.payload;
            const document = state.documents.find((document) => document.name === name);
            if (document) {
                document.states.analyzingStatus = status;
            }
            if (state.currentDocument?.name === name) {
                state.currentDocument!.states.analyzingStatus = status;
            }
        },
        setDocumentLabelingStatus(state, action) {
            const { name, status } = action.payload;
            const document = state.documents.find((document) => document.name === name);
            if (document) {
                document.states.labelingStatus = status;
            }
            if (state.currentDocument?.name === name) {
                state.currentDocument!.states.labelingStatus = status;
            }
        },
        clearCurrentDocument(state) {
            state.currentDocument = null;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(resetPredictions, (state) => {
                state.documents.forEach((document) => {
                    document.states = { ...document.states, analyzingStatus: undefined };
                });
                if (state.currentDocument) {
                    state.currentDocument.states = { ...state.currentDocument.states, analyzingStatus: undefined };
                }
            })
            .addCase(addDocuments.pending, (state, action) => {
                const docs = action.meta.arg;
                docs.forEach((doc) => {
                    state.documents.push({
                        ...doc,
                        thumbnail: "",
                        numPages: 0,
                        currentPage: 0,
                        states: { loadingStatus: DocumentStatus.Loading },
                    });
                });
            })
            .addCase(setCurrentDocument.pending, (state, action) => {
                const document = action.meta.arg;
                const selectedDoc = state.documents.find((doc) => doc.name === document.name);
                if (selectedDoc) {
                    selectedDoc.states.loadingStatus = DocumentStatus.Loading;
                }
            })
            .addCase(setCurrentDocument.fulfilled, (state, action) => {
                const document: IDocument = {
                    ...action.payload.document,
                    states: {
                        ...action.payload.document.states,
                        loadingStatus: DocumentStatus.Loaded,
                    },
                };
                const selectedDoc = state.documents.find((doc) => doc.name === document.name);
                if (selectedDoc) {
                    selectedDoc.states.loadingStatus = DocumentStatus.Loaded;
                }
                state.currentDocument = document;
            })
            .addCase(setCurrentPage.fulfilled, (state, action) => {
                const { pageNumber } = action.payload;
                const document = state.documents.find((doc) => doc.name === state.currentDocument?.name);
                if (document) {
                    document.currentPage = pageNumber;
                }
                if (state.currentDocument) {
                    state.currentDocument.currentPage = pageNumber;
                }
            })
            .addCase(addDocuments.fulfilled, (state, action) => {
                const addedDocs = action.payload;
                const { documents } = state;
                addedDocs.forEach((doc) => {
                    const iDoc = documents.findIndex((d) => d.name === doc.name);
                    if (iDoc >= 0) {
                        documents[iDoc] = {
                            ...doc,
                            states: { ...documents[iDoc].states, loadingStatus: DocumentStatus.Loaded },
                        };
                    }
                });
            });
    },
});

export const { deleteDocument, setDocumentAnalyzingStatus, setDocumentLabelingStatus, clearCurrentDocument } =
    documentsSlice.actions;

export const reducer = documentsSlice.reducer;
