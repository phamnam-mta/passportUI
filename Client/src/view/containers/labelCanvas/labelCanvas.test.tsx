import { shallow } from "enzyme";
import * as React from "react";
import { DocumentStatus, IDocument } from "store/documents/documentsTypes";
import { LabelCanvas } from "./labelCanvas";

describe("<LabelCanvas />", () => {
    let baseProps;
    const currentDocument: IDocument = {
        name: "test",
        type: "pdf",
        url: "<url>",
        thumbnail: "",
        currentPage: 25,
        numPages: 100,
        states: { loadingStatus: DocumentStatus.Loaded },
    };

    beforeEach(() => {
        baseProps = {
            currentDocument: null,
            canvas: {
                imageUrl: "<imageUrl>",
                width: 0,
                height: 0,
                angle: 0,
            },
            onAnalyzeClick: jest.fn(),
            setCurrentPage: jest.fn(),
            setAngle: jest.fn(),
        };
    });

    describe("Rendering", () => {
        it("should render correctly", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when current document is analyzing", () => {
            const props = {
                ...baseProps,
                currentDocument: {
                    ...currentDocument,
                    states: { ...currentDocument.states, analyzingStatus: DocumentStatus.Analyzing },
                },
            };

            const wrapper = shallow(<LabelCanvas {...props} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        const pageControlSelector = "PageControl";
        const imageMapToolbarSelector = "ImageMapToolbar";

        it("should reset progress percentage, when current document is changed", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            wrapper.setProps({ currentDocument });

            expect(wrapper.state("progress")).toBe(undefined);
        });

        it("should go to page, when onPageChange is triggered", () => {
            const props = {
                ...baseProps,
                currentDocument,
            };
            const page = 10;

            const wrapper = shallow(<LabelCanvas {...props} />);
            const pageControl = wrapper.find(pageControlSelector) as any;
            pageControl.prop("onPageChange")(page);

            expect(props.setCurrentPage).toBeCalledTimes(1);
            expect(props.setCurrentPage).toBeCalledWith(page);
        });

        it("should go to previous page, when onPreviousClick is triggered", () => {
            const props = {
                ...baseProps,
                currentDocument,
            };

            const wrapper = shallow(<LabelCanvas {...props} />);
            const pageControl = wrapper.find(pageControlSelector) as any;
            pageControl.prop("onPreviousClick")();

            expect(props.setCurrentPage).toBeCalledTimes(1);
            expect(props.setCurrentPage).toBeCalledWith(currentDocument.currentPage - 1);
        });

        it("should go to next page, when onNextClick is triggered", () => {
            const props = {
                ...baseProps,
                currentDocument,
            };

            const wrapper = shallow(<LabelCanvas {...props} />);
            const pageControl = wrapper.find(pageControlSelector) as any;
            pageControl.prop("onNextClick")();

            expect(props.setCurrentPage).toBeCalledTimes(1);
            expect(props.setCurrentPage).toBeCalledWith(currentDocument.currentPage + 1);
        });

        it("should not go to previous page, when onPreviousClick is triggered and there's no currentDocument", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            const pageControl = wrapper.find(pageControlSelector) as any;
            pageControl.prop("onPreviousClick")();

            expect(baseProps.setCurrentPage).not.toBeCalled();
        });

        it("should not go to next page, when onNextClick is triggered and there's no currentDocument", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            const pageControl = wrapper.find(pageControlSelector) as any;
            pageControl.prop("onNextClick")();

            expect(baseProps.setCurrentPage).not.toBeCalled();
        });

        it("should zoom in, when zoom-in button is clicked", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            const toolbar = wrapper.find(imageMapToolbarSelector) as any;
            wrapper.instance()["imageMap"] = { zoomIn: jest.fn(), getZoom: jest.fn() };
            toolbar.prop("onZoomInClick")();

            expect(wrapper.instance()["imageMap"].zoomIn).toBeCalledTimes(1);
        });

        it("should zoom out, when zoom-out button is clicked", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            const toolbar = wrapper.find(imageMapToolbarSelector) as any;
            wrapper.instance()["imageMap"] = { zoomOut: jest.fn(), getZoom: jest.fn() };
            toolbar.prop("onZoomOutClick")();

            expect(wrapper.instance()["imageMap"].zoomOut).toBeCalledTimes(1);
        });

        it("should zoom to fit, when zoom-to-fit button is clicked", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            const toolbar = wrapper.find(imageMapToolbarSelector) as any;
            wrapper.instance()["imageMap"] = {
                resetZoom: jest.fn(),
                resetCenter: jest.fn(),
            };
            toolbar.prop("onZoomToFitClick")();

            expect(wrapper.instance()["imageMap"].resetZoom).toBeCalledTimes(1);
            expect(wrapper.instance()["imageMap"].resetCenter).toBeCalledTimes(1);
        });

        it("should rotate image clockwise, when rotate button is clicked", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            const toolbar = wrapper.find(imageMapToolbarSelector) as any;
            toolbar.prop("onRotateClick")();

            expect(baseProps.setAngle).toBeCalledTimes(1);
            expect(baseProps.setAngle).toBeCalledWith(90);
        });

        it("should change layer filter states oppositely, when a filter item is clicked", () => {
            const mockItem = { key: "text", checked: true };

            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            const filter = wrapper.find("LayerFilter") as any;
            filter.prop("onItemClick")(mockItem);

            expect(wrapper.state("layerCheckStates")).toMatchObject({ text: false });
        });

        it("should change draw region mode, when Region is clicked", () => {
            const wrapper = shallow(<LabelCanvas {...baseProps} />);
            expect(wrapper.state("drawRegionMode")).toBe(false);

            const button = wrapper.find("DrawRegionButton") as any;
            button.props().onClick();

            expect(wrapper.state("drawRegionMode")).toBe(true);
        });
    });
});
