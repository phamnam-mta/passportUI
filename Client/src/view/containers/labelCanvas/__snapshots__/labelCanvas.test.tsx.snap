// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LabelCanvas /> Rendering should match snapshot, when current document is analyzing 1`] = `
<div
  className="label-canvas"
>
  <div
    className="label-canvas-command-bar"
  >
    <DrawRegionButton
      checked={false}
      disabled={true}
      onClick={[Function]}
    />
    <LayerFilter
      checkStates={
        Object {
          "selectionMarks": true,
          "tables": true,
          "text": true,
        }
      }
      disabled={true}
      onItemClick={[Function]}
    />
  </div>
  <div
    className="label-canvas-image-map"
  >
    <Connect(WithCustomModelLabel)
      currentDocument={
        Object {
          "currentPage": 25,
          "name": "test",
          "numPages": 100,
          "states": Object {
            "analyzingStatus": "Analyzing",
            "loadingStatus": "Loaded",
          },
          "thumbnail": "",
          "type": "pdf",
          "url": "<url>",
        }
      }
      drawRegionMode={false}
      featureStyler={[Function]}
      imageAngle={0}
      imageHeight={0}
      imageUri="<imageUrl>"
      imageWidth={0}
      initLayoutMap={true}
      onMapReady={[Function]}
      setImageMap={[Function]}
    />
    <div
      className="label-canvas-overlay"
    >
      <AnalyzeProgressBar
        subtitle="test"
        title="Running analysis:"
      />
    </div>
  </div>
  <div
    className="label-canvas-control-bar"
  >
    <div
      className="label-canvas-page-control"
    >
      <PageControl
        currentPage={25}
        disabled={true}
        numPages={100}
        onNextClick={[Function]}
        onPageChange={[Function]}
        onPreviousClick={[Function]}
      />
    </div>
    <div
      className="label-canvas-tool-bar"
    >
      <ImageMapToolbar
        disabled={true}
        onRotateClick={[Function]}
        onZoomInClick={[Function]}
        onZoomOutClick={[Function]}
        onZoomToFitClick={[Function]}
        rotateAngle={0}
      />
    </div>
  </div>
</div>
`;

exports[`<LabelCanvas /> Rendering should render correctly 1`] = `
<div
  className="label-canvas"
>
  <div
    className="label-canvas-command-bar"
  >
    <DrawRegionButton
      checked={false}
      disabled={true}
      onClick={[Function]}
    />
    <LayerFilter
      checkStates={
        Object {
          "selectionMarks": true,
          "tables": true,
          "text": true,
        }
      }
      disabled={true}
      onItemClick={[Function]}
    />
  </div>
  <div
    className="label-canvas-image-map"
  >
    <Connect(WithCustomModelLabel)
      currentDocument={null}
      drawRegionMode={false}
      featureStyler={[Function]}
      imageAngle={0}
      imageHeight={0}
      imageUri="<imageUrl>"
      imageWidth={0}
      initLayoutMap={true}
      onMapReady={[Function]}
      setImageMap={[Function]}
    />
  </div>
  <div
    className="label-canvas-control-bar"
  >
    <div
      className="label-canvas-page-control"
    >
      <PageControl
        disabled={true}
        onNextClick={[Function]}
        onPageChange={[Function]}
        onPreviousClick={[Function]}
      />
    </div>
    <div
      className="label-canvas-tool-bar"
    >
      <ImageMapToolbar
        disabled={true}
        onRotateClick={[Function]}
        onZoomInClick={[Function]}
        onZoomOutClick={[Function]}
        onZoomToFitClick={[Function]}
        rotateAngle={0}
      />
    </div>
  </div>
</div>
`;
